DATAWriter, velocity unit conversion, unit test

khuston opened the pull request:
Adds feature in issue #763 

### DATAWriter
Writes selection at current trajectory frame to file, including sections
Atoms, Masses, Velocities, Bonds, Angles, Dihedrals, and Impropers (if
these are defined). Atoms section is written in the "full" sub-style if
charges are available or "molecular" sub-style if they are not.
Molecule id in atoms section is set to to 0.

No other sections are written to the DATA file.
As of this writing, other sections are not parsed into the topology
by the `DATAReader`.

If the selection includes a partial fragment, then the outputted DATA
file will be invalid, because it will describe bonds between atoms
which do not exist in the Atoms section.

By default the writer assumes "conventional" or "real" LAMMPS units
where length is measured in Angstroms and velocity is measured in
Angstroms per femtosecond. If other units are desired, they must be
specified.

Raises ValueError if atom types are not convertable to integers or if
atoms of the same type don't all have the same mass.

An AttributeError will be raised if the atom group doesn't have masses.

### Tests
I added tests to make sure topology attributes of the written file
(after being read again) match the topology attributes read from
the original file. Checks types, bonds, positions, velocities, etc.

There was a problem with test data mini.data because it had atom type
3 which had undefined mass, so I changed the atom type to 1.

### DATAParser change
The DATAParser before was not converting velocity units, so I added
default units Angstroms/fs which are the "real" units in LAMMPS.

PR Checklist
------------
 - [X] Tests?
 - [x] Docs?
 - [ ] CHANGELOG updated?
 - [X] Issue raised/referenced?

richardjgowers commented:
re: the issue with fragments, is this that if you write OH from a water molecule, it will be wrong because it will include the HOH angle?  If so, there's a method called `atomgroup_intersection` on TopologyGroup which does a set intersection between the atomgroup bonds and the topologygroup contents.  So to get bonds that are **only** within the AG:

`ag.bonds.atomgroup_intersection(ag, strict=True)`

Where `strict=True` has made the criteria only match bonds that are entirely within the ag (ie don't extend outside).

khuston commented:
Yes, that's it.

Nice, I didn't know about that method -- that behavior makes more sense than creating a broken file that LAMMPS won't accept.

khuston commented:
On my previous pull request, I changed the atom types dtype from int to '|S1' so select_atoms would work (it wasn't working because the atom type was an int, compared to a string), but I didn't appreciate that this fixes the size of the string in memory to one character, so if a LAMMPS data file had more than 9 atom types, the type would've been truncated at the first digit. I changed the dtype from '|S1' to '|S5' (allowing up to 99999 atom types in a LAMMPS data file... I assume that's more than enough)

I modified a data test file, adding a few atoms to increase to 10 atoms types, and I modified topology/test_lammpsdata.py to check that the number of atom types is as expected. I also cleaned up the tests for DATAWriter a bit.

richardjgowers commented on package/MDAnalysis/coordinates/LAMMPS.py:
```diff
@@ -165,3 +176,212 @@ def _read_first_frame(self):
+        self.file.write('%f %f xlo xhi\n'%(0., triv[0][0]))
+        self.file.write('%f %f ylo yhi\n'%(0., triv[1][1]))
+        self.file.write('%f %f zlo zhi\n'%(0., triv[2][2]))
+        if any(np.array([triv[1][0], triv[2][0], triv[2][1]]) != 0):
```
So anything that isn't 0 is True, so this could boil down to `if any([triv[1][0], triv[2][0], triv[2][1]]):`, no need to make the array either

khuston replied:
Good catch

richardjgowers commented on testsuite/MDAnalysisTests/coordinates/reference.py:
```diff
@@ -200,10 +200,10 @@ class RefTRZ(object):
 class RefLAMMPSData(object):
     filename = LAMMPSdata
-    n_atoms = 18360
+    n_atoms = 18364
```
There's definitely 18360 in the system

richardjgowers replied:
Ohhh right, I just read further down, ignore me.

richardjgowers commented on package/MDAnalysis/coordinates/LAMMPS.py:
```diff
@@ -165,3 +176,212 @@ def _read_first_frame(self):
+    "full" sub-style if charges are available or "molecular" sub-style
+    if they are not. Molecule id is set to 0 for all atoms.
+
+    .. Note::
```
http://sphinxcontrib-napoleon.readthedocs.org/en/latest/example_numpy.html

We're trying to move doc strings to this style instead, it looks better in ipython and the source

richardjgowers commented on package/MDAnalysis/coordinates/LAMMPS.py:
```diff
@@ -165,3 +176,212 @@ def _read_first_frame(self):
+        if has_charges:
+            for index, atype, charge, coords in zip(indices, types, charges,
+                    coordinates):
+                self.file.write('%d 0 %d %f %f %f %f\n'%(index, atype,
```
We're trying to write strings in `'{a:d} {b:f}'.format(a=this, b=that)` style

richardjgowers commented on package/MDAnalysis/coordinates/LAMMPS.py:
```diff
@@ -165,3 +176,212 @@ def _read_first_frame(self):
+        self.units = {'time': 'fs', 'length': 'Angstrom'}
+        self.units['length'] = kwargs.pop('lengthunit', self.units['length'])
+        self.units['time'] = kwargs.pop('timeunit', self.units['time'])
+        self.units['velocity'] = kwargs.pop('velocityunit', \
```
You don't need the backslash for line continuations, just make the bracket line up [like here](https://www.python.org/dev/peps/pep-0008/#maximum-line-length)

richardjgowers commented:
Looking good, just style fixes for now.  Bit annoying but it keeps everything uniform for later on.

khuston commented:
I decided to leave out an atom renumbering option since that would be achievable with `Merge` when it's working again.

kain88-de commented:
@khuston have you made progress on this?

khuston commented:
Unless there are suggested changes, or recent changes to the 363 branch have introduced new errors, I think it's ready to be merged. A few notes:

1. I have used this for my own project with Lammps data and dcd files for a bit and haven't run into problems.
2. image indices are not preserved, if the user reads a lammps data file with wrapped positions and image indices
3. I thought the following code worked:

    u = mda.Universe('1R2M.pdb')
    ad = {}
    for i, j in enumerate(set(u.atoms.types)):
        ad[j] = i
    u.atoms.types = np.array([ad[i] for i in u.atoms.types])
    u.add_TopologyAttr(mda.core.topologyattrs.Masses([1 for x in u.atoms]))
    
but apparently it gives this error:

    TypeError: only integer arrays with one element can be converted to an index

I'm not sure what's happening here. Actually I should probably check whether this is a problem with the writer itself, or with some other part of the code.

khuston commented:
The `TypeError` I just referenced only arises because I passed a list to `add_TopologyAttr` instead of an array. However, I also discovered another problem which occurs if two different atom types share the same mass when I'm generating the type-to-mass dictionary for the data file.

richardjgowers commented:
Yeah attrs will only work with np arrays because we use fancy slicing. Maybe we should put in an asarray call to the init.

Not sure I understand the problem with masses?

khuston commented:
@richardjgowers Never mind the masses, I misidentified the problem. I think this is the simplest code that will create a lammps data file from a pdb file _without_ bonds.

    u = mda.Universe(pdb)
    ad = {j:i for i,j in enumerate(set(u.atoms.types))}
    # Re-assign atom types as string integers b/c lammps requires integer types, but
    # select_atoms will fail if atoms.types array actually has dtype int instead of str
    u.add_TopologyAttr(mda.core.topologyattrs.Atomtypes(np.array([ad[i] for i in u.atoms.types],
                                                                 dtype='S')))
    # give masses to atoms
    u.add_TopologyAttr(mda.core.topologyattrs.Masses(12.*np.ones(len(u.atoms))))
    with mda.Writer('output.data') as W:
        W.write(u.atoms)
    
However, if there are bonds, the data file will have a bad Bonds section:

    Bonds

    1 ('2', '2') 18 377
    2 ('2', '2') 90 317
    ...

I could either index the bond types, or leave out bonds section if bond types are not explicitly defined. What do you think?

jdetle commented:
@khuston, I tried to checkout your DATAWriter branch to play with it, why is log.py gone? I'm getting issues with MDAnalysis initialization as a result.

khuston commented:
@jdetle Which log.py are you referring to? It's not in the list of files changed. I see:

- package/MDAnalysis/core/log.py
- package/MDAnalysis/lib/log.py

jdetle commented:
It's probably an issue with my git skills because I switch from my branch to yours and I no longer have `package/MDAnalysis/lib/log.py` OR `package/MDAnalysis/core/log.py`.

khuston commented:
Weird, I get both of them when I clone my repository and checkout the DATAWriter branch.

jdetle commented:
If you have any tips for helping me out, the steps I took were:

git remote add khuston httpsgithub.com/khuston/MDAnalysis
git fetch khuston 
git checkout khuston/DATAWriter
git checkout -b DATAWriter

khuston commented:
@richardjgowers I found another issue. Requiring that atom types are integers is insufficient to write a valid lammps data file. Given N total atom types, the only acceptable atom types are 1, ... N. This leads to an annoying problem.

Given a universe of atoms of 2 types (1 and 2), all atoms of type 2 are selected. This selection is then saved to a data file. There are two valid ways to save the data file:
Approach A. Include both atom types in the header and `Masses` section, even though only atoms of type 2 are defined in the `Atoms` section. Do this by including data from all types in the universe, if possible. Otherwise, make up masses for the unknown types.
Approach B. Re-number atom types, changing type 2 to type 1 for writing the new data file.

Approach A only works the first time the selection is written. If a data file defining 2 atom types but only including instances of one of the types is read, then the writer won't know about the absent type the second time around, because there are no atoms of type 1 in the universe. A hack-ish workaround is then to write fake masses (e.g. 0 or 1) for as many types as needed. For example, if the user is writing atoms of type 7 to a data file, and there are no other types in the universe, the resulting `Masses` section would look like

    Masses
    
    1 1.0
    2 1.0
    ...
    6 1.0
    7 72.0
    
    Atoms
    (atoms of type 7)

where only the known mass of type 7 is real, and the others are made up. 

For my own purposes, I find this scheme of preserving atom type indices (Approach A) to be most convenient. However, this is elaborate behavior to get around lammps/vmd topotools restrictions on atom types + MDA not recording attributes of atoms that aren't part of the universe, and users could want different behavior.

khuston commented:
Since there's been no comment for several weeks, I pushed DATAWriter with the following rules:

- Writer raises an exception if there are bonds but bond types are not _explicitly defined_ as integers >= 1, i.e. will raise an exception given the default output of `bond.type`.
- Atomtypes must be integers >= 1.
- Atomtypes are not renumbered on writing.
- The Masses section is written for all atomtypes up to the maximum atomtype represented in the written selection. If the mass for an atomtype cannot be determined from the selection's universe, then it is taken to be 1.

These rules make DATAWriter strict, but it also ensures lammps will accept the written files, which I assume is the only reason someone would want to write a lammps data file.

For the conflicts, I think issue-363 snippets can be all accepted (and DATAWriter snippets rejected) except for the imports in the lammps test.

richardjgowers commented:
Hi Kyle, sorry I thought I replied to this.  I think I agree, it makes sense that any written data file must be valid, rather than representing a slice of the original data file.  I'll look this over soon(tm) and try and get it merged.

richardjgowers commented:
@khuston merged, thanks!
